Add Hybrid Rendering post (#518)
withastro/astro.build: src/content/blog/hybrid-rendering.mdx

// Images
import staticBuild from '/src/assets/blog/hybrid-rendering/static.png'
import serverBuild from '/src/assets/blog/hybrid-rendering/server.png'
import hybridBuild from '/src/assets/blog/hybrid-rendering/hybrid.png'

Since the initial launch of Astro 1.0, developers have been able to choose between generating [static (SSG)](https://docs.astro.build/en/core-concepts/routing/#static-ssg-mode) or [server (SSR)](https://docs.astro.build/en/guides/server-side-rendering/) output. Static websites offer incredible performance by building your pages ahead-of-time. Servers leverage their dynamic power to generate HTML on-demand, unique for every request.

Until now, each Astro project has required an all-or-nothing decision—is this fully static or should I deploy a server?

[Astro 2.0](/astro-2) unlocks the best of both worlds with **Hybrid Rendering**. Choose between the performance of static pages or the full flexibility of server rendering for each page of your website.

In this post, we'll go over how Hybrid Rendering works by exploring just a few of the new features and performance improvements made possible in Astro 2.0.

## How does `astro build` work?

> We'll simplify things a bit for clarity, but we first need to understand that Astro's build process happens in multiple stages. First, Astro uses [Vite](https://vitejs.dev) to generate a server-side JavaScript bundle. While this bundle is being generated, we perform static analysis on your source code to generate a **client manifest**. The output of this bundle includes:

Reviewer: Also there’s a slight disruption to the flow of this paragraph because it goes:

1. “generate a bundle”
2. “while bundle, client manifest”
3. “bundle includes”

Could it make sense to remove the client manifest mention here and just add a mention of “static analysis” to its bullet below?

The bolding of “client manifest” particularly threw me because it makes it seem like the most important thing in that paragraph but at this point I have no idea what it is.

Author: Agreed! I removed the bolding and will definitely rephrase this one.

-   **Server-side JavaScript** used to render HTML

> -   A **client manifest** which marks any components needed for client-side interactivity

Reviewer: Is there a better word than “marks” here? And does the manifest “mark” or does some other process “mark” by putting the component in the manifest? 

Speaking from ignorance so don’t know the best wording but here’s a random suggestion:

```suggestion
-   A **client manifest** which collects all components needed for client-side interactivity
```

Next, Astro uses the **client manifest** to initiate a second build process which bundles optimized client-side JavaScript.

If your configured [`output`](https://docs.astro.build/en/reference/configuration-reference/#output) is `static`, Astro will execute the server-side JavaScript and write the output to `.html` files. The server-side JavaScript is then discarded.

<BlogContentImage
    src={staticBuild}
    alt="Flow diagram of Astro's static build process"
/>

If your configured [`output`](https://docs.astro.build/en/reference/configuration-reference/#output) is `server`, Astro passes the server-side JavaScript to an [`adapter`](https://docs.astro.build/en/reference/configuration-reference/#adapter) for further processing.
Adapters ensure that your server-side JavaScript is compatible with a particular hosting provider's JavaScript runtime. Note that in this case, the final output is not a set of `.html` files, but the JavaScript code necessary to render HTML.

<BlogContentImage
    src={serverBuild}
    alt="Flow diagram of Astro's server build process"
/>

## How does hybrid rendering work?

In Astro 2.0, we've revamped the `astro build` pipeline to handle hybrid rendering.

> During the initial bundling process, we've added an additional static analysis step that determines which pages should be prerendered.

Reviewer: Suggestion to avoid “we’ve” repeated across sentences and slightly punchier sentence:

```suggestion
During the initial bundling process, a new static analysis step determines which pages should be prerendered.
```

/>

## Why use static analysis?

Static analysis is a technique used to analyze your source code without actually executing it.
Astro relies on this approach in many places, but it is especially critical during the build process.
For performance reasons, Astro must be able to categorize your pages without executing their source code.
For smaller projects, execution might not be a concern, but at scale it quickly becomes a bottleneck.

Hybrid rendering was specifically designed with static analysis in mind, though we did [entertain alternative APIs](https://github.com/withastro/roadmap/blob/main/proposals/0029-prerender-api.md#alternatives).
Astro's build process is able to check for the presence of an `export const prerender = true` statement to determine which pages should be prerendered.
To do this efficiently, we rely on the wonderful [`es-module-lexer`](https://github.com/guybedford/es-module-lexer) library, which is also used internally in Node.

```astro
---
// This route should be generated at build time!
export const prerender = true

const text = await fetch('https://example.com/').then((res) => res.text())
---

<article set:html={text} />
```

> ## Use cases for hybrid rendering

Reviewer 2: Subheadings are rarely useful in a blog post IMO, since someone is reading it top-to-bottom they aren't organizing the information in the same way as docs.

I would push away from the "sub-headings" here and go back to the three `##` headings that all start with `## Use case:` or `## Case Study:`

Reviewer 2: It's okay if the sections are short, although the second one "Add an API" could use a second paragraph, even if it  a bit of filler.

> Hybrid Rendering unlocks a whole new set of possibile use cases, but we wanted to highlight a few that we're particularly excited about.

Reviewer 3: ```suggestion
Hybrid Rendering unlocks a whole new set of possible use cases, but we wanted to highlight a few that we're particularly excited about.
```

Hybrid rendering solves our performance problem by rendering the homepage to static HTML ahead-of-time, while keeping the rest of your website rendering on-demand.

### Add an API to your existing static site

As your static website grows, you may find the need to add a public API. Historically, adding an API to a static site has required host-specific configuration, but in Astro 2.0, we can add [Server Endpoints (API Routes)](https://docs.astro.build/en/core-concepts/endpoints/#server-endpoints-api-routes) without sacrificing the performance of our static pages.

### Improve build performance of large sites

Static generation can be difficult to scale to thousands of pages, since every page must be rendered during the build process. Server-side rendering defers page rendering until request-time, removing a potential bottleneck from your build process.

When there are hundreds of routes generated by `getStaticPaths`, server-side rendering can lead to significant performance gains. In our internal testing, we saw build times improve by up to 30% by switching dynamic routes to server-side rendering.

## Getting started with hybrid rendering

> Hybrid rendering is available today in Astro 2.0! You can head to our [server-side rendering guide](https://docs.astro.build/en/guides/server-side-rendering/) to learn more about Astro's SSR features. Astro's official adapters have also been updated to support hybrid rendering, so be sure to install the latest version of your [adapter](https://docs.astro.build/en/guides/server-side-rendering/#adding-an-adapter).

Reviewer 2: can we add a reference to https://www.youtube.com/watch?v=AFP3ZHxO7Gg ? Happy to help wordsmith but would be great to help them promote it more
